extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduRangeRead.java

```diff
@@ -1,189 +1,266 @@
-            KuduDataIndexRow.deserializeDataIndexRow(r, visibilityEnabled));
+    }, new RowConsumer(results));
+  }
+
+  public void checkFinalize(
+      final AsyncKuduScanner scanner,
+      final Semaphore semaphore,
+      final BlockingQueue<Object> resultQueue,
+      final AtomicInteger queryCount) {
+    scanner.close();
+    semaphore.release();
+    if (queryCount.decrementAndGet() <= 0) {
+      try {
+        resultQueue.put(RowConsumer.POISON);
+      } catch (final InterruptedException e) {
+        LOGGER.error("Interrupted while finishing blocking queue, this may result in deadlock!");
       }
-      tmpIterator =
-          Arrays.stream(dataIds).map(d -> resultsMap.get(new ByteArray(d))).filter(
-              Objects::nonNull).iterator();
-      return new CloseableIteratorWrapper<>(() -> {
-      }, (Iterator<T>) tmpIterator);
     }
   }
 
+  public Deferred<Object> executeScanner(
+      final AsyncKuduScanner scanner,
+      final Semaphore semaphore,
+      final BlockingQueue<Object> resultQueue,
+      final AtomicInteger queryCount,
+      final AtomicBoolean isCanceled,
+      final boolean visibilityEnabled,
+      final Predicate<GeoWaveRow> filter,
+      final GeoWaveRowIteratorTransformer<T> rowTransformer,
+      final boolean rowMerging,
+      final short[] adapterIds) {
+    // Errback class
+    class QueryErrback implements Callback<Deferred<Object>, Exception> {
+      @Override
+      public Deferred<Object> call(final Exception e) {
+        LOGGER.warn("While scanning rows from kudu", e);
+        checkFinalize(scanner, semaphore, resultQueue, queryCount);
+        return Deferred.fromError(e);
+      }
+    }
+
+    // callback class
+    class QueryCallback implements Callback<Deferred<Object>, RowResultIterator> {
+      @Override
+      public Deferred<Object> call(RowResultIterator rs) {
+        Stream<GeoWaveRow> tmpStream;
+
+        if (rs == null || isCanceled.get()) {
+          checkFinalize(scanner, semaphore, resultQueue, queryCount);
+          return Deferred.fromResult(null);
+        }
+
+        if (rs.getNumRows() > 0) {
+          tmpStream = Streams.stream(rs.iterator()).map(r -> (GeoWaveRow) new KuduRow(r));
+          if (visibilityEnabled) {
+            tmpStream = tmpStream.filter(filter);
+          }
+
+          Iterator<GeoWaveRow> tmpIterator = tmpStream.iterator();
+
+          rowTransformer.apply(
+              rowMerging ? new GeoWaveRowMergingIterator(tmpIterator)
+                  : tmpIterator).forEachRemaining(row -> {
+                    try {
+                      resultQueue.put(row);
+                    } catch (final InterruptedException e) {
+                      LOGGER.warn("interrupted while waiting to enqueue a kudu result", e);
+                    }
+                  });
+
+        }
+
+        if (scanner.hasMoreRows()) {
+          return scanner.nextRows().addCallbackDeferring(this).addErrback(new QueryErrback());
```

Comment:
Could we make this `QueryErrback` instance static, so we don't have to construct a new one for every row?